### ISSUE TITLE
Remove and/or propagate experimental API usages

### ISSUE THREAD

**Issue by martinbonnin** [apollographql/apollo-android#3395]
`dev-3.x` uses opt-in annotations such as `@OptIn(ExperimentalCoroutinesApi::class)` and `@OptIn(ExperimentalContracts::class)`. We should either add a `ApolloExperimental` or remove these before going stable to make sure they do not break compatibility.

**Comment by BoD** [comment]
## TLDR

- remove 2 instances of `ExperimentalUnsignedTypes`, and `ExperimentalCoroutinesApi` on `ApolloClient.executeAsFlow` (no impact)
- mark `OkHttpWebSocketEngine` private
- remove `DelicateCoroutinesApi` on `runTest` (`runWithMainLoopApple` and `runWithMainLoopJs`): update code to not use `GlobalScope`
- add `ApolloExperimental` on `MainLoopDispatcher` in `appleMain`
- `WebSocketNetworkTransport.execute`: not sure what to do!

## More in-depth

Here's a list of all places where `@OptIn` is used and the proposed action to take:

|What                                                                           |Where                            |Why                                                                                                     |What to do                  |More info                                                                                                  |
|-------------------------------------------------------------------------------|---------------------------------|--------------------------------------------------------------------------------------------------------|----------------------|-----------------------------------------------------------------------------------------------------------|
|@OptIn(ApolloInternal::class)                                                  |CustomScalarAdapters             | Version2CustomTypeAdapterToAdapter needs to be public but is marked as internal so clients don't use it|Keep                  |There on purpose                                                                                           |
|@OptIn(ApolloInternal::class)                                                  |ApolloClient.Builder             | Version2CustomTypeAdapterToAdapter needs to be public but is marked as internal so clients don't use it|Keep                  |There on purpose                                                                                           |
|@OptIn(ExperimentalContracts::class)                                           |JsonWriters                      |Use of contract                                                                                         |Keep                  |On internal API                                                                                            |
|@OptIn(ExperimentalTime::class)                                                |CodegenTest                      |Use of Duration to monitor generation times                                                             |Keep    |In Tests. (Also, I think it's stable in Kotlin 1.6)                                                                                                   |
|@OptIn(ExperimentalStdlibApi::class)                                           |CodegenTest                      |Use of buildList                                                                                        |Keep   |In Tests (Also, I think it's stable in Kotlin 1.6)                                                                                                   |
|@OptIn(ApolloExperimental::class)                                              |DefaultApolloExtension           |generateTestBuilders is experimental                                                                    |Keep                  |There on purpose                                                                                           |
|@OptIn(ApolloExperimental::class)                                              |TestBuildersTest                 |Use of generated Test Builder code                                                                      |Keep                  |There on purpose                                                                                           |
|@OptIn(ExperimentalUnsignedTypes::class)                                       |NativeMockServer                 |?                                                                                                       |Remove                |No warning when removing                                                                                   |
|@OptIn(ExperimentalUnsignedTypes::class)                                       |NSURLSessionHttpEngine           |?                                                                                                       |Remove                |No warning when removing                                                                                   |
|@OptIn(InternalCoroutinesApi::class)                                           |DefaultDispatcher in appleMain   |Referring to Delay                                                                                      |Keep                  |On internal API                                                                                            |
|@OptIn(InternalCoroutinesApi::class)                                           |MainLoopDispatcher in appleMain  |Referring to Delay                                                                                      |Add ApolloExperimental|Used only in 1 place                                                                                       |
|@OptIn(ExperimentalCoroutinesApi::class)                                       |DefaultDispatcher in appleMain   |Use of resumeUndispatched                                                                               |Keep                  |On internal API                                                                                            |
|@OptIn(ExperimentalCoroutinesApi::class)                                       |NSURLSessionWebSocketEngine      |Use of invokeOnClose                                                                                    |Keep                  |On private API                                                                                             |
|@OptIn(ExperimentalCoroutinesApi::class, kotlinx.coroutines.FlowPreview::class)|ApolloClient.executeAsFlow       |?                                                                                                       |Remove                |No warning when removing                                                                                   |
|@OptIn(ExperimentalCoroutinesApi::class)                                       |WebSocketNetworkTransport.execute|Use of transformWhile                                                                                   |Not sure              |Public API + on an interface override. Maybe we could refactor to not need takeWhile but that may be touchy|
|@OptIn(ExperimentalCoroutinesApi::class)                                       |OkHttpWebSocketEngine            |Use of invokeOnClose                                                                                    |Keep                  |This could/should be private?                                                                              |
|@OptIn(DelicateCoroutinesApi::class)                                           |runWithMainLoopApple.runTest     |Use of GlobalScope                                                                                      |Remove                |Use a dedicated CoroutineScope instead?                                                                    |
|@OptIn(DelicateCoroutinesApi::class)                                           |runWithMainLoopJs.runTest        |Use of GlobalScope                                                                                      |Remove                |Use a dedicated CoroutineScope instead?                                                                    |


## Action plan
- If an OptIn is on internal or private API, or tests: keep as-is
- If an OptIn can be removed with no impact, remove
- If an OptIn is on a public API
  - if the API should actually be private, make it private ;)
  - if OptIn can be removed by tweaking the code, do it and remove
  - not a prominent API: add `ApolloExperimental` (clients will need to opt in)
  - prominent API + not easy to tweak: not sure! (it's the case of `WebSocketNetworkTransport.execute`)

**Comment by martinbonnin** [comment]
Early feedbacks, will dive more into this:

> DelicateCoroutinesApi

I think this is ok to `@Opt-in` `DelicateCoroutinesApi` as we "know what we're doing" 😅 

> If an OptIn is on internal or private API

Unfortunately, I don't think we can do that. If `kotlinx.coroutines` removes an experimental API in a future version, all users will have start having runtime crashes, even if it's used in a private Apollo API.

Overall, I think it's either:
1. We can remove it completely (best case)
2. We can't remove it. Then either propagate it to the user if it's an advanced API. If not, workaround it somehow

In the `WebSocketNetworkTransport.execute` case, could we copy/paste the coroutines `transformWhile` implementation in Apollo? Maybe it's not too much code?

**Comment by BoD** [comment]
Agreed!  I've updated the comment above taking this into consideration.  I think the only "don't know what to do" remaining is Apple's `DefaultDispatcher` relying on `Delay`.

**Comment by BoD** [comment]
Handled in #3565
